Ray.jl/src/runtime.jl

```diff
@@ -173,23 +173,76 @@ end
 
 initialize_coreworker_driver(args...) = ray_jll.initialize_coreworker_driver(args...)
 
+# TODO: Move task related code into a "task.jl" file
 function submit_task(f::Function, args::Tuple, kwargs::NamedTuple=NamedTuple();
                      runtime_env::Union{RuntimeEnv,Nothing}=nothing,
                      resources::Dict{String,Float64}=Dict("CPU" => 1.0))
     export_function!(FUNCTION_MANAGER[], f, get_current_job_id())
     fd = ray_jll.function_descriptor(f)
-    arg_oids = map(Ray.put, flatten_args(args, kwargs))
+    task_args = serialize_args(flatten_args(args, kwargs))
 
     serialized_runtime_env_info = if !isnothing(runtime_env)
         _serialize(RuntimeEnvInfo(runtime_env))
     else
         ""
     end
 
-    return GC.@preserve args ray_jll._submit_task(fd,
-                                                  arg_oids,
-                                                  serialized_runtime_env_info,
-                                                  resources)
+    GC.@preserve task_args begin
+        return ray_jll._submit_task(fd,
+                                    transform_task_args(task_args),
+                                    serialized_runtime_env_info,
+                                    resources)
+    end
+end
+
+# Adapted from `prepare_args_internal`:
+# https://github.com/ray-project/ray/blob/ray-2.5.1/python/ray/_raylet.pyx#L673
+function serialize_args(args)
+    ray_config = ray_jll.RayConfigInstance()
+    put_threshold = ray_jll.max_direct_call_object_size(ray_config)
+    rpc_inline_threshold = ray_jll.task_rpc_inlined_bytes_limit(ray_config)
+    record_call_site = ray_jll.record_ref_creation_sites(ray_config)
+
+    rpc_address = ray_jll.GetRpcAddress()
+
+    total_inlined = 0
+    task_args = Any[]
+    for arg in args
+        # Note: The Python `prepare_args_internal` function checks if the `arg` is an
+        # `ObjectRef` and in that case uses the object ID to directly make a
+        # `TaskArgByReference`. However, as the `args` here are flattened the `arg` will
+        # always be a `Pair` (or a list in Python). I suspect this Python code path just
+        # dead code so we'll exclude it from ours.
```

Comment:
_should_ we be doing this though?  the values will be automatically dereferenced on the the task execution side so I'm not so sure; then again, if we _don't_ do it then users may be surprised if they _do_ have to de-reference in their work functions.

Reply:
The Python code I'm referring to is attempting to avoid creating a new `ObjectRef` here if the user has passed in one as an argument. In all cases Python will dereference the arguments inside the called task. What I was specifically calling out here is that due to `flatten_args` this optimization will never occur as an `arg` is never a `ObjectRef`. At best it would be a `list` containing an `ObjectRef` as the second value.